src/main/java/net/openhft/chronicle/network/cluster/handlers/HeartbeatHandler.java

```diff
@@ -134,12 +136,20 @@ public void onRead(@NotNull WireIn inWire, @NotNull WireOut outWire) {
 
     @Override
     public void close() {
-        if (connectionMonitor != null)
-            connectionMonitor.onDisconnected(localIdentifier(), remoteIdentifier(), nc().isAcceptor());
-        lastTimeMessageReceived = Long.MAX_VALUE;
-        Closeable closable = closable();
-        if (closable != null && !closable.isClosed()) {
-            Closeable.closeQuietly(closable);
+        if (closed.compareAndSet(false, true)) {
+            if (connectionMonitor != null) {
+                try {
+                    connectionMonitor.onDisconnected(localIdentifier(), remoteIdentifier(), nc().isAcceptor());
+                } catch (Exception e) {
+                    Jvm.error().on(getClass(), "close:", e);
+                    throw Jvm.rethrow(e);
+                }
+            }
+            lastTimeMessageReceived = Long.MAX_VALUE;
+            Closeable closable = closable();
+            if (closable != null && !closable.isClosed()) {
+                Closeable.closeQuietly(closable);
+            }
```

Comment:
If we make this close idempotent, and we call onDisconnected in the close method, we only end up calling it once (it always gets closed when disconnecting)